lib/grammers-client/src/types/attributes.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 - developers of the `grammers` project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use grammers_tl_types as tl;
+
+pub enum Attribute {
+    Audio {
+        duration: i32,
+        title: Option<String>,
+        performer: Option<String>,
+    },
+    Voice {
+        duration: i32,
+        waveform: Option<Vec<u8>>,
```

Comment:
This will need some explanation on how waveforms are to be calculated/rendered…

Reply:
Actually, I have no idea how they are calculated  :D

Reply:
Unfortunately, I think the rendering is inconsistent, but at least on Telegram Desktop it's known.